### PR TITLE
WIP DicomClient / DicomService stability improvements (Tests are still sometimes crashing)

### LINKED ISSUE #982: DicomClientSendingRequestsState holds onto the sent requests
### Expected behavior
The requests that have been sent are discarded and eventually garbage collected.

### Actual behavior
After a request is sent, `DicomClientSendingRequestsState` removes it from the queue and puts it to `ConcurrentBag<DicomRequest> _sentRequests`. This causes memory usage to blow up since the request objects cannot be garbage collected. `_sentRequests` seems to be only used for its `Count` property; the requests could certainly be counted with some other means, too.

### Steps to reproduce the behavior
Send a few hundred images in an association and the process will soon consume close to 1 GB of memory that is freed only after the association is released. This can also be seen using Visual Studio's Performance Profiler which shows an ever-increasing count of allocations eventually pointing to request objects that cannot be collected.

### fo-dicom version and OS/platform
fo-dicom 4.0.4

### LINKED ISSUE #1054: 4.0.4 DicomClient stucks on SendAsync() for hours
Dear contributors and supporters, I have the following problem. Thank you in advance for reading!

### Expected behavior
SendAsync() should return (or crash) at least within timeout (set to 2 minutes)

### Actual behavior
About once in a few weeks it happens that SendAsync() never returns - for hours.
The rest of my appl is running and collecting images.

### Steps to reproduce the behavior
quite difficult, very seldom

### fo-dicom version and OS/platform
4.0.4* in .NET 4.8 x86 exe on Windows Server 

*unfortunately I am currently not able to update to 4.0.5 here - If an update to 4.0.5 might solve this, it would be a very interesting information

I am calling the send like:
`try{
// preparation here
await dicomClient.SendAsync();
} catch (exception e)
{
Log("This is not seen when problm happens");
}`
(I am very sorry if I am wrong - or use the library wrong....)

The log shortly before it stucks is:
...
```
12:16:02,833 Trace - DicomLog(): VI-DEMO1 -> C-Store request [22450]
12:16:02,989 Trace - DicomLog(): VI-DEMO1 <- C-Store response [22450]: Success
12:16:02,989 Trace - DicomLog(): [SENDING REQUESTS] All requests are done, going to linger association now...
12:16:02,989 Trace - DicomLog(): [SENDING REQUESTS] --> [LINGERING ASSOCIATION]
12:16:03,052 Trace - DicomLog(): [LINGERING ASSOCIATION] Linger timed out after 50ms, releasing association
12:16:03,052 Trace - DicomLog(): [LINGERING ASSOCIATION] --> [RELEASING ASSOCIATION]
12:16:03,052 Trace - DicomLog(): VI-DEMO1 -> Association release request
12:16:03,052 Trace - DicomLog(): VI-DEMO1 <- Association release response
12:16:03,052 Trace - DicomLog(): Connection closed
12:16:03,052 Trace - DicomLog(): [RELEASING ASSOCIATION] Association release response received, disconnecting...
12:16:03,052 Trace - DicomLog(): [RELEASING ASSOCIATION] --> [COMPLETED]
12:16:03,052 Trace - DicomLog(): [COMPLETED] DICOM client completed without errors
12:16:03,052 Trace - DicomLog(): [COMPLETED] Cleaning up
12:16:03,052 Trace - DicomLog(): [COMPLETED] --> [IDLE]
12:16:03,052 Trace - DicomLog(): [IDLE] No requests to send, staying idle
12:16:03,271 Trace - DicomLog(): [IDLE] --> [CONNECTING]
12:16:03,271 Trace - DicomLog(): [CONNECTING] --> [REQUESTING ASSOCIATION]
12:16:03,271 Trace - DicomLog(): VI-DEMO1 -> Association request:
12:16:03,271 Trace - DicomLog(): VI-DEMO1 <- Association accept:
12:16:03,271 Trace - DicomLog(): [REQUESTING ASSOCIATION] Association is accepted
12:16:03,271 Trace - DicomLog(): [REQUESTING ASSOCIATION] --> [SENDING REQUESTS]
12:16:03,271 Trace - DicomLog(): [SENDING REQUESTS] Sending DICOM requests
12:16:03,271 Trace - DicomLog(): VI-DEMO1 -> C-Store request [22451]
12:16:03,411 Trace - DicomLog(): [SENDING REQUESTS] DICOM client connection still has unsent requests, sending those now...
12:16:03,427 Trace - DicomLog(): VI-DEMO1 <- C-Store response [22451]: Success
12:16:03,443 Trace - DicomLog(): VI-DEMO1 -> C-Store request [22452]
12:16:03,583 Trace - DicomLog(): VI-DEMO1 <- C-Store response [22452]: Success
12:16:03,583 Trace - DicomLog(): VI-DEMO1 -> C-Store request [22453]
12:16:03,724 Trace - DicomLog(): VI-DEMO1 <- C-Store response [22453]: Success
12:16:03,724 Trace - DicomLog(): [SENDING REQUESTS] All requests are done, going to linger association now...
12:16:03,724 Trace - DicomLog(): [SENDING REQUESTS] --> [LINGERING ASSOCIATION]
12:16:03,786 Trace - DicomLog(): [LINGERING ASSOCIATION] Linger timed out after 50ms, releasing association
12:16:03,786 Trace - DicomLog(): [LINGERING ASSOCIATION] --> [RELEASING ASSOCIATION]
12:16:03,786 Trace - DicomLog(): VI-DEMO1 -> Association release request
12:16:03,786 Trace - DicomLog(): VI-DEMO1 <- Association release response
12:16:03,786 Trace - DicomLog(): Connection closed
12:16:03,786 Trace - DicomLog(): [RELEASING ASSOCIATION] Disconnected during association release, cleaning up...
12:16:03,786 Trace - DicomLog(): [RELEASING ASSOCIATION] --> [COMPLETED]
12:16:03,786 Trace - DicomLog(): [COMPLETED] DICOM client completed without errors
12:16:03,786 Trace - DicomLog(): [COMPLETED] Cleaning up
12:16:03,786 Trace - DicomLog(): [COMPLETED] --> [IDLE]
12:16:03,786 Trace - DicomLog(): [IDLE] No requests to send, staying idle
12:16:04,005 Trace - DicomLog(): [IDLE] --> [CONNECTING]
12:16:04,005 Trace - DicomLog(): [CONNECTING] --> [REQUESTING ASSOCIATION]
12:16:04,005 Trace - DicomLog(): VI-DEMO1 -> Association request:
12:16:04,005 Trace - DicomLog(): VI-DEMO1 <- Association accept:
12:16:04,005 Trace - DicomLog(): [REQUESTING ASSOCIATION] Association is accepted
12:16:04,005 Trace - DicomLog(): [REQUESTING ASSOCIATION] --> [SENDING REQUESTS]
12:16:04,005 Trace - DicomLog(): [SENDING REQUESTS] Sending DICOM requests
12:16:04,021 Trace - DicomLog(): VI-DEMO1 -> C-Store request [22460]
12:16:23,130 Trace - DicomLog(): Socket error while reading PDU: TimedOut [10060]
12:16:23,130 Trace - DicomLog(): Socket error while writing PDU: ConnectionReset [10054]
12:16:23,130 Trace - DicomLog(): Connection closed
12:16:23,130 Trace - DicomLog(): Connection closed
12:16:24,068 Warn - DicomLog(): Exception creating PDV: System.IO.IOException: Unable to write data to the transport connection: An existing connection was forcibly closed by the remote host. ---> System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote host
12:16:24,099 Warn - DicomLog(): Exception writing data to PDV: System.IO.IOException: Unable to write data to the transport connection: An existing connection was forcibly closed by the remote host. ---> System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote host
12:16:24,162 Warn - DicomLog(): Exception sending DIMSE: System.IO.IOException: Unable to write data to the transport connection: An existing connection was forcibly closed by the remote host. ---> System.Net.Sockets.SocketException: An existing connection was forcibly closed by the remote host
12:16:24,162 Trace - DicomLog(): VI-DEMO1 -> C-Store request [22461]
```

### LINKED ISSUE #1071: DicomService.cs: recursion in Task CheckForTimeouts() potentially resulting in stack overflow exception
### Expected behavior
DicomService.cs - Task CheckForTimeouts() holds no recursion

### Actual behavior
During the course of the C-MOVE actions the Task CheckForTimeouts is involved (depending on RequestTimeout). Within this task there's recursion (CheckForTimeouts() is called from within Task CheckForTimeouts()), eventually leading to a stackoverflow exception.
The call stack from the thread involved is filled with (DebugDiag and WinDBG): 
`Dicom.Network.DicomService+<CheckForTimeouts>d__67.MoveNext())`

### Steps to reproduce the behavior
trace stack when handling a long running C-MOVE-RQ

### fo-dicom version and OS/platform
4.0.4 (also valid for 4.0.5) .NET 4.5.2

### PR DESCRIPTION
**Pull request by amoerie**
Update: 13/07/2020 Some tests are still failing, but they are unrelated to this PR. Debating how to proceed...

Fixes #1054 
Fixes #982
Fixes #1071 

#### Checklist
- [x] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [x] I have updated the change log
- [x] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Notify DicomClient of AssociationRelease when the connection is closed before we receive an AssociationReleased event. This can happen when the server closes the connection very quickly.
- Ensure the DicomClientSendingRequestsState._sending flag is set back to false, even if something goes wrong while sending a request
- Ensure that the DicomService._pending queue is updated when a request goes awry.
- Force close the connection when sending a PDU over the network fails. Otherwise, the message remains stuck in the pending queue and the DicomClient will infinitely keep trying to send it.
- Improve DicomServer disposal by immediately disposing all of its services when the server gets disposed. 
- Add test that simulates a socket exception from the server by disposing the DICOM server halfway between sending requests
- Add test that simulates a socket exception from the client by throwing a custom IOException while writing to the networkstream
- Add test that simulates a generic (non-socket) exception from the client by throwing a custom Exception while writing to the networkstream. I have confirmed that this test causes the DicomClient to infinitely keep trying to send a DICOM request. In fact, this behavior is part of DicomService, so it could also show up as a DicomServer that is stuck trying to send a DICOM request infinitely. 
- I've updated the Github Actions to run the tests for each target framework (net462, netcoreapp2.1 and netcoreapp3.1) in parallel
- Use a count instead of a list of references to the sent DicomRequests to reduce memory pressure while sending many/large DICOM requests
- Fix multithreading issue in parallel transcoder test
- Do not use recursion when looping to check timed out requests, this can cause Stack Overflow errors.
- Do not use recursion when looping to keep sending DICOM requests from within the DicomClient, this can cause Stack Overflow errors.

### CODE REVIEW

**Comment by gofal** [comment]
I updated the title of the pull request with a "WIP" title, so that I do not merge this PR by accident like I did with other pull requests 🙄

**Comment by amoerie** [comment]
Tests are green on my machine. And now we wait. 🕐

**Comment by amoerie** [comment]
@gofal ready when you are 👍🙂

File: DICOM/Network/IDicomServiceRunner.cs [hunk 1]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the Microsoft Public License (MS-PL).
 
 #if !NET35
+using System.Threading;
```
Comment (gofal) [inline DICOM/Network/IDicomServiceRunner.cs:5]: this using is not necessary

**Comment by gofal** [comment]
now one test failed: FellowOakDicom.Tests.Network.Client.DicomClientTest.SendAsync_Linger_ShouldKeepDelayingLingerAsLongAsRequestsAreComingIn(numberOfRequests: 2, secondsBetweenEachRequest: 4, lingerTimeoutInSeconds: 5) [FAIL]

**Comment by amoerie** [comment]
🤦‍♂️ great. I'll have to review that one. It's a test for the old DicomClient that is a little tricky to get right because the SendAsync of the old DicomClient completes way too early, so you have to jump through hoops to "wait until the requests are processed".

I'll take a look first thing in the morning. Thanks for reviewing!

**Comment by gofal** [comment]
thanks! Its not an easy job you have choosen to dig into the DicomClient.

**Comment by amoerie** [comment]
Look at that! I managed to get the 4.x release line building in Github Actions! 🤩

File: DICOM/Network/IDicomServiceRunner.cs (hunk 1 above)
Comment (amoerie) [reply DICOM/Network/IDicomServiceRunner.cs]: nice catch

**Comment by amoerie** [comment]
It would seem that the "random" test host crashes are related to DicomTranscoderTranscode_ToCompressedCodecInParallel_NoMultithreadIssues. Investigation is pending...

**Comment by amoerie** [comment]
@gofal I found an issue in GH549.cs where the transcoders are run in a multithreaded fashion. 

More specifically, the test seems to run a specific transcoder on a list of DICOM datasets in parallel, and checks that no exceptions are thrown. 

Unfortunately, this test was the cause of of the test host crashing from time to time (see CI failures of this PR)

This is the implementation, as of right now in development:

https://github.com/fo-dicom/fo-dicom/blob/f22da83962f255dd3f8369d6d3bf10a462e34e59/Tests/FO-DICOM.Tests/Bugs/GH549.cs#L19-L47

The problematic line, it seems to me, is the following:
https://github.com/fo-dicom/fo-dicom/blob/f22da83962f255dd3f8369d6d3bf10a462e34e59/Tests/FO-DICOM.Tests/Bugs/GH549.cs#L25

You see, Enumerable.Repeat will create an enumerable, where each instance is a reference to a single clone. The question we need to ask ourselves is, do we want to run the transcoders all on the same DICOM data set, or do we want to run them on different data sets?

I've replaced the aforementioned line with the following: 

https://github.com/fo-dicom/fo-dicom/blob/d7afc490605f8ae694e2d9a363b68edd0695aa7d/Tests/FO-DICOM.Tests/Bugs/GH549.cs#L25

Now, it will create a clone of the DICOM dataset and pass that into the transcoders.

**Comment by amoerie** [comment]
@gofal  (sorry for pinging you so much lately)

I am now getting a failing test, again in the rendering using ImageManager. What do I do? It looks like the tests are not very robust yet in our v5 release pipeline, and these errors seem unrelated to this PR. Do we merge the PR and open another one to look at the CI issues? It feels a bit dishonest to say that this PR is red, while in fact the PR about the issue templates is also red because of flaky unit tests.

This is the log of the failing test now:

```log
2020-07-13T09:48:25.8910855Z [xUnit.net 00:00:02.00]     FellowOakDicom.Tests.Imaging.DicomImageTest.ManipulatedImage_MultithreadedAccess_ShouldNotThrow [FAIL]
2020-07-13T09:48:25.8911162Z [INFO] ANY-SCP -> C-Echo request [20]
2020-07-13T09:48:25.8916999Z   X FellowOakDicom.Tests.Imaging.DicomImageTest.ManipulatedImage_MultithreadedAccess_ShouldNotThrow [550ms]
2020-07-13T09:48:25.8918236Z   Error Message:
2020-07-13T09:48:25.8918751Z    Assert.Null() Failure
2020-07-13T09:48:25.8920025Z Expected: (null)
2020-07-13T09:48:25.8921643Z Actual:   System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object.) ---> System.NullReferenceException: Object reference not set to an instance of an object.
2020-07-13T09:48:25.8922320Z    at FellowOakDicom.Imaging.Render.ImageGraphic.RenderImage(ILUT lut) in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Imaging\Render\ImageGraphic.cs:line 231
2020-07-13T09:48:25.8922689Z    at FellowOakDicom.Imaging.DicomImage.RenderImage(Int32 frame) in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Imaging\DicomImage.cs:line 265
2020-07-13T09:48:25.8923070Z    at FellowOakDicom.Tests.Imaging.DicomImageTest.<>c__DisplayClass2_0.<ManipulatedImage_MultithreadedAccess_ShouldNotThrow>b__1(Int32 i) in D:\a\fo-dicom\fo-dicom\Tests\FO-DICOM.Tests\Imaging\DicomImageTest.cs:line 41
2020-07-13T09:48:25.8923434Z    at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
2020-07-13T09:48:25.8923876Z --- End of stack trace from previous location where exception was thrown ---
2020-07-13T09:48:25.8924214Z    at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
2020-07-13T09:48:25.8924537Z    at System.Threading.Tasks.TaskReplicator.Replica`1.ExecuteAction(Boolean& yieldedBeforeCompletion)
2020-07-13T09:48:25.8924852Z    at System.Threading.Tasks.TaskReplicator.Replica.Execute()
2020-07-13T09:48:25.8925144Z    --- End of inner exception stack trace ---
2020-07-13T09:48:25.8925418Z    at System.Threading.Tasks.TaskReplicator.Run[TState](ReplicatableUserAction`1 action, ParallelOptions options, Boolean stopOnFirstFailure)
2020-07-13T09:48:25.8925901Z    at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
2020-07-13T09:48:25.8926304Z --- End of stack trace from previous location where exception was thrown ---
2020-07-13T09:48:25.8926645Z    at System.Threading.Tasks.Parallel.ThrowSingleCancellationExceptionOrOtherException(ICollection exceptions, CancellationToken cancelToken, Exception otherException)
2020-07-13T09:48:25.8927003Z    at System.Threading.Tasks.Parallel.ForWorker[TLocal](Int32 fromInclusive, Int32 toExclusive, ParallelOptions parallelOptions, Action`1 body, Action`2 bodyWithState, Func`4 bodyWithLocal, Func`1 localInit, Action`1 localFinally)
2020-07-13T09:48:25.8927379Z    at System.Threading.Tasks.Parallel.For(Int32 fromInclusive, Int32 toExclusive, Action`1 body)
2020-07-13T09:48:25.8927697Z    at FellowOakDicom.Tests.Imaging.DicomImageTest.<>c__DisplayClass2_0.<ManipulatedImage_MultithreadedAccess_ShouldNotThrow>b__0() in D:\a\fo-dicom\fo-dicom\Tests\FO-DICOM.Tests\Imaging\DicomImageTest.cs:line 44
2020-07-13T09:48:25.8928050Z    at Xunit.Record.Exception(Action testCode) in C:\Dev\xunit\xunit\src\xunit.core\Record.cs:line 26
2020-07-13T09:48:25.8928429Z ---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object.
2020-07-13T09:48:25.8928739Z    at FellowOakDicom.Imaging.Render.ImageGraphic.RenderImage(ILUT lut) in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Imaging\Render\ImageGraphic.cs:line 231
2020-07-13T09:48:25.8929059Z    at FellowOakDicom.Imaging.DicomImage.RenderImage(Int32 frame) in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Imaging\DicomImage.cs:line 265
2020-07-13T09:48:25.8930108Z    at FellowOakDicom.Tests.Imaging.DicomImageTest.<>c__DisplayClass2_0.<ManipulatedImage_MultithreadedAccess_ShouldNotThrow>b__1(Int32 i) in D:\a\fo-dicom\fo-dicom\Tests\FO-DICOM.Tests\Imaging\DicomImageTest.cs:line 41
2020-07-13T09:48:25.8930485Z    at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
2020-07-13T09:48:25.8930910Z --- End of stack trace from previous location where exception was thrown ---
2020-07-13T09:48:25.8931571Z    at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
2020-07-13T09:48:25.8931899Z    at System.Threading.Tasks.TaskReplicator.Replica`1.ExecuteAction(Boolean& yieldedBeforeCompletion)
2020-07-13T09:48:25.8932218Z    at System.Threading.Tasks.TaskReplicator.Replica.Execute()<---
2020-07-13T09:48:25.8932452Z 
2020-07-13T09:48:25.8932684Z   Stack Trace:
2020-07-13T09:48:25.8933417Z      at FellowOakDicom.Tests.Imaging.DicomImageTest.ManipulatedImage_MultithreadedAccess_ShouldNotThrow() in D:\a\fo-dicom\fo-dicom\Tests\FO-DICOM.Tests\Imaging\DicomImageTest.cs:line 34

```

**Comment by gofal** [comment]
Yes, agree that the flickering unittests may be related to codec. Because they always had been green until I integrated the codec.
These tests had been completely disabled until now because there was no codec available at all. So if the issue is within the codec code, or if the unitttests which are calling the codec, are bad, would be an interesting question.
Great you managed to isolate a unittests whose crash caused some random other tests to fail. I already saw that you posted it on Efferents issues. Thanks for that!

**Comment by gofal** [comment]
by the way: unitl now I never noticed any unexplainable red tests on my local machine. it all seems only to happen on CI. So some things like parallelism, timeout due to slower instances, or similar must be the root cause.

**Comment by amoerie** [comment]
I've created https://github.com/fo-dicom/fo-dicom/issues/1072 to track this CI issue. 

My proposal would be to merge this PR, because it includes a bunch of important fixes, and figure out the remaining CI issues separately. What do you think @gofal ?

**Comment by gofal** [comment]
Yes, sure. please merge this.

Do you think, this is fine enough to do an 4.0.6 release?

**Comment by amoerie** [comment]
Yes, I think the improvements in this PR and others would be a good 4.0.6 candidate. The network tests are consistently green, although I must admit I haven't done extensive testing in the field yet with these changes.

**Pull request merged**